### DISCUSSION TITLE
关于异步日志在生产环境下的适用性

### DISCUSSION THREAD

**Discussion by yinghaoyu** [chenshuo/muduo#633] [Ideas]
背景：在轨道交通或者航空领域，研发人员是没有权限在现场对程序debug的，取日志就需要层层审批，非常麻烦。

Muduo的设计在两处可能会缺失日志：
1. `FileUtil`使用`setbuffer`设置了本地缓冲区，假如这个时候buffer没有及时写入磁盘，程序直接dump了，会缺失部分日志。
2. 同理在`AsyncLogging`内，日志写入太频繁的时候也会drop部分日志。

缺失日志，对bug分析并不友好。
请陈硕哥分享一下对这个问题的看法。

大家有想法的，也可以分享一下，谢谢诸君！

**Comment by chenshuo** [comment]
首先，diagnostic logs = transaction logs. muduo 提供的是前者，如果你需要后者，请使用 SQLite.
其次，如果程序 core dump，而你又能拿到 core file，可以通过查找 cookieStart 来找到没有及时写入磁盘的 log buffer.
最后，日志写入速度受限于你给程序分配的磁盘带宽，如果程序生成日志的速度大于可用磁盘带宽，要么丢日志，要么 out-of-memory。

**Reply by HelloWorld-9527** [reply]
你好，作者先生，这里说到的

> 可以通过查找 cookieStart 来找到没有及时写入磁盘的 log buffer.

看代码中 cookie_ 保存着 cookieStart ，然后是怎么来找到没有及时写入的log_buffer的呢
